pkg/image/containersimageregistry/registry.go

```diff
@@ -0,0 +1,290 @@
+
+func New(sourceCtx *types.SystemContext, opts ...Option) (orimage.Registry, error) {
+	reg := &Registry{
+		sourceCtx: sourceCtx,
+	}
+
+	for _, opt := range opts {
+		if err := opt(reg); err != nil {
+			return nil, err
+		}
+	}
+
+	if reg.cache == nil {
+		var err error
+		reg.cache, err = getDefaultImageCache()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return reg, nil
+}
+
+func NewDefault() (orimage.Registry, error) {
+	return New(DefaultSystemContext)
+}
+
+type cacheConfig struct {
+	baseDir  string
+	preserve bool
+}
+
+func (c *cacheConfig) ociLayoutDir() string {
+	return filepath.Join(c.baseDir, "oci-layout")
+}
+func (c *cacheConfig) blobInfoCacheDir() string {
+	return filepath.Join(c.baseDir, "blob-info-cache")
+}
+
+func (c *cacheConfig) getSystemContext() *types.SystemContext {
+	return &types.SystemContext{
+		BlobInfoCacheDir: c.blobInfoCacheDir(),
+	}
+}
+
+type Option func(*Registry) error
+
+func getDefaultImageCache() (*cacheConfig, error) {
+	if dir := os.Getenv("OLM_CACHE_DIR"); dir != "" {
+		return newCacheConfig(filepath.Join(dir, "images"), true), nil
+	}
+	return getTemporaryImageCache()
+}
+
+func getTemporaryImageCache() (*cacheConfig, error) {
+	tmpDir, err := os.MkdirTemp("", "opm-containers-image-cache-")
+	if err != nil {
+		return nil, err
+	}
+	return newCacheConfig(tmpDir, false), nil
+}
+
+func newCacheConfig(dir string, preserve bool) *cacheConfig {
+	return &cacheConfig{
+		baseDir:  dir,
+		preserve: preserve,
+	}
+}
+
+func ForceTemporaryImageCache() Option {
+	return func(r *Registry) error {
+		var err error
+		r.cache, err = getTemporaryImageCache()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func InsecureSkipTLSVerify(insecureSkipTLSVerify bool) Option {
+	return func(r *Registry) error {
+		if r.sourceCtx == nil {
+			r.sourceCtx = &types.SystemContext{}
+		}
+		r.sourceCtx.DockerDaemonInsecureSkipTLSVerify = insecureSkipTLSVerify
+		r.sourceCtx.DockerInsecureSkipTLSVerify = types.NewOptionalBool(insecureSkipTLSVerify)
+		r.sourceCtx.OCIInsecureSkipTLSVerify = insecureSkipTLSVerify
+		return nil
+	}
+}
+
+func (r *Registry) Pull(ctx context.Context, ref orimage.Reference) error {
+	namedRef, err := reference.ParseNamed(ref.String())
+	if err != nil {
+		return err
+	}
+	dockerRef, err := docker.NewReference(namedRef)
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(r.cache.ociLayoutDir(), 0700); err != nil {
+		return err
+	}
+	ociLayoutRef, err := layout.NewReference(r.cache.ociLayoutDir(), ref.String())
+	if err != nil {
+		return err
+	}
+
+	policy, err := signature.DefaultPolicy(r.sourceCtx)
+	if err != nil {
+		return err
+	}
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return err
+	}
+
+	sourceCtx := r.sourceCtx
+	authFile := getAuthFile(r.sourceCtx, namedRef.String())
+	if authFile != "" {
+		if sourceCtx == nil {
+			sourceCtx = &types.SystemContext{}
```

Comment:
use default?

Reply:
No, with pretty much the same reasoning as above. At this point, the caller has already explicitly provided us with a `nil` SystemContext and has opted _not_ to use the `DefaultSystemContext`.